### ISSUE TITLE
Add twistedchecker to Travis

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8572]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8572|
|Type|enhancement|
|Created|2016-07-08 09:28:41Z|
|Branch|https://github.com/twisted/twisted/tree/8572-txchecker-travis|

There's no real reason to run twistedchecker on the buildbots: Travis is quite capable of doing it.

<details><summary>Searchable metadata</summary>

```
trac-id__8572 8572
type__enhancement enhancement
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8572_txchecker_travis 8572-txchecker-travis
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1467970121590110 1467970121590110
changetime__1470518956870931 1470518956870931
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban set owner to @adiroiban|
|-|-|

Let me give it a try :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

https://github.com/twisted/twisted/pull/315

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban removed owner|
|-|-|

I have created a (Linux/Unix only for now) script at `bin/admin/twistedchecker-trunk-diff.sh`

It can take an argument an run twistedchecker with diff on it.

It was created for Travis, but can also be used on Linux/OSX.

By taking an argument you can run the checks on a subset of the files, resulting in faster runs on the local branch.

So if you have only made changes in twisted/words you can run the checks only for twisted/words.

How to test it:


```
# Run for an unknown dir
$ bin/admin/twistedchecker-trunk-diff.sh no-such-dir
no-such-dir does not exists as a path or as a module.
$ echo $?
1

# Run for an unknown module
$ bin/admin/twistedchecker-trunk-diff.sh twisted.missing
no-such-dir does not exists as a path or as a module.
$ echo $?
1


# Run for an path
$ bin/admin/twistedchecker-trunk-diff.sh twisted/words
$ echo $?
0

```


To trigger an error apply this diff

```
diff --git twisted/words/__init__.py twisted/words/__init__.py
index 8b05bd0..eb1b29d 100644
--- twisted/words/__init__.py
+++ twisted/words/__init__.py
@@ -1,4 +1,3 @@
-# -*- test-case-name: twisted.words.test -*-
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
```

then run


```
$ bin/admin/twistedchecker-trunk-diff.sh twisted.words
************* Module twisted.words
W9002:  1,0: Missing a reference to test module in header
$ echo $?
1
```


You can see a failing tests here

https://travis-ci.org/twisted/twisted/builds/143764569

Note that for testing I have disabled all the other travis-ci tests to get quicker results.

For now, the twistedchecker is crammed together with the other py2.7 tests. I hope that [#8535](https://github.com/twisted/twisted/issues/8535) will be accepted and when we can have a dedicated job for the static checkers

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks for constantly working to improve the CI situation, Adi.

I think that this should be in a travis-specific script for the time being and not a generalized dev tool.  If we want to generalize it, we should add this functionality to twistedchecker itself (which is probably a good idea, but out of scope for this ticket).  On mimic we have a `.travis/` directory with some shell scripts in it and I would suggest repeating that pattern here.

More importantly, there are conflicts that need resolving, and there are unrelated changes made to the travis config that should probably be in a separate change.

I've also left some comments in the PR about the implementation strategy.

Please address this feedback and resubmit.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban removed owner|
|-|-|

Hi,

Thanks for the review.

----------

I have moved the script in .travis but this does not stop me from using it on my Linux dev system on a daily basis :)

`.travis/twistedchecker-trunk-diff.sh twisted/conch/ssh/` is much faster than using buildbot and I can get some results without a commit

-------

by leaving it in bin/admin other developers can discover the script and if they are using Windows they might want to write an OS agnostic version of that script :)


Please check latest changes.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @adiroiban|
|-|-|

There are merge conflicts to be resolved, but this is a good start.  Please land it.

In the future, we should:

1. rewrite the git shenanigans as a Python script so it's more portable
2. perhaps add the git shenanigans to twistedchecker itself for use on other projects
3. remove the twistedchecker buildbot - the only thing that's between this PR and a green checkmark on travis!
4. separate out the twistedchecker build into a separate travis build-matrix element, so we can see it separately.

but as I said; this is a good start, there's no reason not to land it and improve from there.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

Thanks for the review.

1 and 2) I think that the git stuff should be moved to git.

3) Once this is merged I will remove the twistedchecker buildbot builder.

4) Please see [#8535](https://github.com/twisted/twisted/issues/8535)

waiting for tests results and will merge.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|GitHub set status to `closed`|
|-|-|

In changeset 8a58bb3aca81cad8f3e8075ae57174a49257d153
```
#!CommitTicketReference repository="" revision="8a58bb3aca81cad8f3e8075ae57174a49257d153"
Merge 8572-txchecker-travis: Add twistedchecker to Travis-CI jobs.


Author: adiroiban
Reviewer: glyph
Fixes: #8572
```

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_9"></a>|@rodrigc commented|
|-|-|

This isn't working very well.
twistedchecker is throwing exceptions while it runs, and this is flagging the entire build as being broken:

https://travis-ci.org/twisted/twisted/jobs/146015829

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

Thanks for the report. I will revert.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

In changeset 7c5cfe3656618b0535449ee99bd52d911506d835
```
#!CommitTicketReference repository="" revision="7c5cfe3656618b0535449ee99bd52d911506d835"
Revert "Merge 8572-txchecker-travis: Add twistedchecker to Travis-CI jobs."

twistedchecker --dif get false positives and it is now required.

Reopens: #8572
```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set status to `reopened`|
|-|-|

Looks like the `reopens` hook is not working

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_13"></a>|@rodrigc commented|
|-|-|

This should be re-done, but in combination with  [ticket:8535]

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban commented|
|-|-|

well, first we need to fix twistedchecker --diff  - https://github.com/twisted/twistedchecker/issues/115

since twistedchecker buildbot builder was already broken I will not revert the buildbot changes. Let me know if you want twistedchecker back on buildbot

To not block this ticket on twistedchecker issues, we can introduce the twistedchecker travis-ci job as an allow_failure job.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban removed owner<br>@adiroiban set status to `new`|
|-|-|

Ok. I went for using the diff_cover project.

Looks like the preferred tool for pylint/coverage ... etc

Rather than reinventing the wheel, I think that is better to work together with other people interested in diff_cover [1] to provide a pylint base diff tool... as twistedchecker is just pylint :)

You can see a travis-ci run here: https://travis-ci.org/twisted/twisted/jobs/148667038

For now the checker is in the "allow failure" group so that in the next weeks we can monitor its behaviour.

The script is still in .sh format. For a follow up ticket maybe we can re-write it as python so that we can use it on Windows.

----------------

The diff semantic is different, rather than looking for newly introduces errors, the current check looks for errors triggered by the diff lines

This mean that if you move bad code, you will need to fix the moved code.... but previous twistedcheck-diff had the same problem.


-------------

On Linux/OSX you can re-use the travis helper command to run twistechecker.

```

# Whole project
$ .travis/twistedchecker-trunk-diff.sh twisted
# A specific module/package
$ .travis/twistedchecker-trunk-diff.sh twisted.words.service
```


Check that line offset will not trigger false positives

```
diff --git a/twisted/words/service.py b/twisted/words/service.py
index d3ac900..85a7320 100644
--- a/twisted/words/service.py
+++ b/twisted/words/service.py
@@ -25,6 +25,7 @@ How does this thing work?
 
   - protocol specific hangup, realm is notified, user is removed from active
     play, the end.
+
 """
 
 from time import time, ctime
```


If the diff touches a line which has failures it will fail... even if the line already triggers an error in trunk:

```
diff --git a/twisted/words/service.py b/twisted/words/service.py
index d3ac900..1509232 100644
--- a/twisted/words/service.py
+++ b/twisted/words/service.py
@@ -65,7 +65,7 @@ class Group(object):
 
 
     def add(self, user):
-        assert iwords.IChatClient.providedBy(user), "%r is not a chat client" % (user,)
+        assert iwords.IChatClient.providedBy(user), "%r is not a chat client " % (user,)
         if user.name not in self.users:
             additions = []
             self.users[user.name] = user
```


---------

Please check if is a good idea to merge the current version of twistedchecker

Thanks!

* [1] https://github.com/Bachmann1234/diff-cover

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks for this, adi: I'm really happy to see progress on this.  There are a few issues though:

1. the build you linked, https://travis-ci.org/twisted/twisted/jobs/148667038, contains a bunch of Python tracebacks, and there's no message or anything indicating that those are expected.
2. There are sadly some conflicts that need resolving, so this needs to be updated with trunk.

If you can deal with those tracebacks (either actually fix them, or just echo a message indicating to the user that they're expected, and a link to a bug report tracking their fix), update the conflicts, and get a successful travis run, then please go ahead and merge this as an experimental build; I like the idea of running something like this as an allow_fail first, that is a good idea; it also massively reduces the quality standard to merge it :).

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_17"></a>|@adiroiban commented|
|-|-|

The traceback are from twistedchecker and not related to these changes.

I can no longer access the buildbot twistedchecker logs, but I remember that we were getting the same errors... is just that nobody was trying to filter them among the huge amount of other errors that were reported

I have created this ticket https://github.com/twisted/twistedchecker/issues/118 together with printing a message in travis-ci job

-----------

Starting with a builder/job as allow failure is similar to what we were doing for unsupported platforms.

Now with Travis-CI and App-veyour the build results for allow_failures are a bit harder to observer in a PR as the global result is green... but I am considering a solution

I think that having a builder in allow failure should not reduce the quality required to merge a change.


Waiting for build results and will merge

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|GitHub set status to `closed`|
|-|-|

In changeset 8ee02e19add51d85066fb0cd61bfc119519dba1e
```
#!CommitTicketReference repository="" revision="8ee02e19add51d85066fb0cd61bfc119519dba1e"
Merge 8572-txchecker-travis: Add twistedchecker to Travis-CI jobs.


Author: adiroiban
Reviewer: glyph
Fixes: #8572
```
